go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll

```diff
@@ -623,13 +623,20 @@
       not cmpl.isNormal()
     }
 
-    predicate succ(ControlFlow::Node pred, ControlFlow::Node succ) {
+    /**
+     * Holds if `succ` is a successor of `pred`, ignoring the execution of any
+     * deferred functions when a function ends.
+     */
+    predicate succ0(ControlFlow::Node pred, ControlFlow::Node succ) {
```

Comment:
Hah! This is a correct warning. The magic optimisation restricts a predicate to only be computed where some use-site might use the result -- for example, if I have `private predicate evens(int x) { x % 2 = 0 and x >= 0 and x <= 1000000 } and the only user is `... x <= 10 and evens(x)`, magic might introduce the side-condition `x <= 10` into `evens` and so make computing it cheaper. If that is applied to `succ0` it could at least in theory restrict succ0 to (among other things) defer sequences defined in terms of succ0, and thus reintroduce recursion, though the optimiser should try pretty hard not to create recursion via magic.

Long story short, make this `pragma[nomagic] private`.

Reply:
Sure. This alert is for the member predicate of `ControlFlowTree`. Do I need to do it for the overrides as well? And should I also do it for the top-level predicate?

Reply:
Good question; not sure how the pragma interacts with overrides (and for that matter, not sure about the combination of private and overrides). Suggest posing the Q to the ql channel.

Reply:
I've asked to be sure, but I think it doesn't get inherited. I've added a commit to add the pragma to all the member predicates. (They can't be private, btw, but they're on a private class.) I'm still unsure about the top level predicate, on line 2115. (Also not sure if it should still be cached, since it's now private and only used to build the top level `succ`.)